### PR TITLE
[ MB- 15032 ] Store origin duty location gbloc on orders table

### PR DESCRIPTION
**Pull request by kctruss**
## [Jira ticket](https://dp3.atlassian.net/browse/MB-15032) for this change

This ticket ended up involving a good deal of server test tinkering and refactor as it fundamentally changes GBLOC expectations. I cut down the usage of `testdatagen.MakePostalCodeToGBLOC()` except where it explicitly made sense to keep, and found a couple of cases (so far) of circular data creation (i.e. `testdatagen.MakeMove` => `testdatagen.MakeOrders` => `link move to orders`) - This pattern is not necessary as `MakeMove` already generates an associated orders record. At this time, server tests pass on my local environment, but please verify if testing changes locally due to the extent of changes required to cover this work.

## Summary

[ADR #77 ](https://transcom.github.io/mymove-docs/docs/adrs/permanently-store-origin-duty-gbloc/) provides context and reasoning behind the approach used and the need for permanently storing GBLOC values.

## Setup to Run Your Code

<details>
<summary>💻 You will need to use two separate terminals to test this locally.</summary>

##### Terminal 1

Start the UI locally.

```sh
make client_run
```

##### Terminal 2

Start the Go server locally.

```sh
make server_run
```

</details>

### Additional steps

1. Access the office app
2. Login as a multi-account user
3. Verify that moves populate in the associated queues (SC, TOO, TIO, Prime) as expected based on GBLOC.

~~NOTE: USMC will be stored as USMC on the orders table and routes to the USMC GBLOC queues, but existing logic in the API still selects a GBLOC to display based on the postal code of the origin address of the first shipment on the order in the queue list. This is currently expected behavior.~~ This is confirmed to not be an expectation, was initially an assumption on my part but [that has been cleared up](https://ustcdp3.slack.com/archives/C04LXDA81PV/p1678306815928949)

## Verification Steps for Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [x] Request review from a member of a different team.
- [ ] Have the Jira acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/dev/contributing/backend/Backend-Programming-Guide#logging)
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/dev/contributing/backend/Backend-Programming-Guide/#querying-the-database-safely) have been satisfied.

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/dev/contributing/database/Database-Migrations#zero-downtime-migrations)
- [ ] Have been communicated to #g-database
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/dev/contributing/database/Database-Migrations#secure-migrations)

### CODE REVIEW

File: pkg/handlers/internalapi/orders.go [hunk 1]
```diff
@@ -139,6 +139,8 @@ func (h CreateOrdersHandler) Handle(params ordersop.CreateOrdersParams) middlewa
 				return handlers.ResponseForError(appCtx.Logger(), err), err
 			}
 			originDutyLocation := serviceMember.DutyLocation
+			originDutyLocationGBLOC := serviceMember.DutyLocation.TransportationOffice.Gbloc
```
Comment (duncan-truss) [inline pkg/handlers/internalapi/orders.go:142]: How does this work when we have a non-installation duty location like `Houston, TX 77037` that doesn't have a Transportation Office associated with it?

Comment (kctruss) [reply pkg/handlers/internalapi/orders.go:142]: That would be a case I missed. I'll update this.

File: pkg/services/order/order_updater.go [hunk 2]
```diff
@@ -544,6 +544,10 @@ func updateOrderInTx(appCtx appcontext.AppContext, order models.Order, checks ..
 		order.OriginDutyLocationID = &originDutyLocation.ID
 		order.OriginDutyLocation = &originDutyLocation
 
+		// if originDutyLocation.TransportationOffice.Gbloc != "" {
```
Comment (kctruss) [inline pkg/services/order/order_updater.go:547]: I did not mean to leave this commented out. Working on a related fix.

**Review by ronaktruss — COMMENTED** [review]
I likely won't have a chance to re-review this, but Katy and I spent some time looking at this PR and talked about what changes are expected before this PR is merged in. They are the following:
1. Always using the postal code look up to determine gbloc
2. Remove usages of TransportationOffice.Gbloc in this context (as a result of change #1)
3. Update tests accordingly
4. determine if we need to support updating the gbloc when the order is updated (tentatively it doesn't look like we need to support this as we don't have a way to update the orders gbloc after the initial creation)

@kctruss Let me know if you think there is anything I misrepresented. To whoever is reviewing this feel free to ask about why those decisions were made if they aren't clear (they likely aren't).

**Comment by duncan-truss** [comment]
> I likely won't have a chance to re-review this, but Katy and I spent some time looking at this PR and talked about what changes are expected before this PR is merged in. They are the following:
> 
>     1. Always using the postal code look up to determine gbloc
> 
>     2. Remove usages of TransportationOffice.Gbloc in this context (as a result of change #1)
> 
>     3. Update tests accordingly
> 
>     4. determine if we need to support updating the orders.gbloc when the service member is updated (it tentatively doesn't look like we will have the orders created when this is possible)
> 
> 
> @kctruss Let me know if you think there is anything I misrepresented. To whoever is reviewing this feel free to ask about why those decisions were made if they aren't clear (they likely aren't).

I am pretty sure we'd need to support the customer going back and editing their current duty station with the PATCH /service-member endpoint from the review page.

On the office side both the SC and the TOO can edit the origin duty station on the orders.  On an SC this would potentially remove it from their queue.  For a TOO it may only have an effect on routing when it's an NTS shipment without an address that relies on the origin duty location info.

**Comment by kctruss** [comment]
In implementing the above discussed changes/missed requirements, I'm running into a new spider web of server test failures that have been really frustrating and are holding this work up. I'm moving this PR back into draft and releasing EXP for the time being. I'll communicate the state of the work in Slack @ronaktruss so you and Jules have a sense of where it's at.

**Comment by kctruss** [comment]
> I am pretty sure we'd need to support the customer going back and editing their current duty station with the PATCH /service-member endpoint from the review page after the orders have been created.
> 
> On the office side both the SC and the TOO can edit the origin duty station on the orders. On an SC this would potentially remove it from their queue. For a TOO it may only have an effect on routing when it's an NTS shipment without an address that relies on the origin duty location info.

Confirmed that behavior is maintained with my latest changes.

**Comment by duncan-truss** [comment]
@kctruss I wasn't able to get the GBLOC update when I edited the service member's current duty location.  I think I would expect to see updates to this handler code where we update other order fields when that is modified

https://github.com/transcom/mymove/blob/43864b2bd2bc2e5aa07f7a1b9a203c63f0862e28/pkg/handlers/internalapi/service_members.go#L220-L223

Outside of that I can confirm updating the orders is changing the GBLOC value

File: swagger-def/internal.yaml [hunk 3]
```diff
@@ -1828,6 +1828,9 @@ definitions:
       origin_duty_location:
         $ref: 'definitions/DutyLocationPayload.yaml'
         x-nullable: true
+      origin_duty_location_gbloc:
+        type: string
+        title: From what GBLOC do your orders originate?
```
Comment (ronaktruss) [inline swagger-def/internal.yaml:1833]: Should this be a description? Given the existing usages I think this is fine as is. 

Screenshot 2023-03-28 at 11 04 54 AM

**Review by ronaktruss — COMMENTED** [review]
Just reading through the code, nothing caught my eye. So this generally looks good to me.

**Review by duncan-truss — APPROVED** [review]
Thanks for figuring out the service member update part 🚢

File: .circleci/config.yml [hunk 4]
```diff
@@ -36,7 +36,7 @@ references:
   dp3-branch: &dp3-branch placeholder_branch_name
   # MUST BE ONE OF: loadtest, demo, exp.
   # These are used to pull in env vars so the spelling matters!
-  dp3-env: &dp3-env placeholder_env
+  dp3-env: &dp3-env exp
```
Comment (duncan-truss) [inline .circleci/config.yml:39]: Reminder to revert this

File: migrations/app/schema/20230223225426_store_origin_duty_gbloc.up.sql [hunk 5]
```diff
@@ -0,0 +1,22 @@
+-- Restore the gbloc column with the previous settings
+ALTER TABLE orders
+	ADD COLUMN gbloc VARCHAR;
+
+CREATE INDEX orders_gbloc_idx ON orders (gbloc);
```
Comment (duncan-truss) [inline migrations/app/schema/20230223225426_store_origin_duty_gbloc.up.sql:5]: It's usually faster to apply the index after a bunch of updates or insertions but not too worried about system impacts here.

**Pull request merged**